File: gateway/router.py
from __future__ import annotations

"""Routing strategies.

- round_robin     : naive load balancer, cache-blind (the baseline to beat).
- consistent_hash : deterministic affinity on the FIRST block only. No
                    longest-prefix match, no eviction awareness -> collapses
                    families that share a system prompt onto one backend.
- prefix_tree     : longest-prefix match via the radix tree, scored by estimated
                    TTFT (affinity + load) with saturation cutoff + hysteresis.
"""

from .config import Config
from .extensions.bpe_hashing import block_hashes_with_fallback
from .extensions.ttft_predictor import TTFTPredictor, blended_ttft, features_for
from .hashing import block_hashes
from .load_tracker import LoadTracker
from .radix_tree import RadixTree


class RouteResult:
    __slots__ = ("backend_id", "hashes", "tokens", "match_blocks", "hash_mode")

    def __init__(self, backend_id, hashes, tokens, match_blocks, hash_mode="char"):
        self.backend_id = backend_id
        self.hashes = hashes
        self.tokens = tokens
        self.match_blocks = match_blocks
        self.hash_mode = hash_mode


class Router:
    def __init__(self, cfg: Config, tree: RadixTree, load: LoadTracker,
                 predictor: TTFTPredictor | None = None, fleet=None):
        self.cfg = cfg
        self.tree = tree
        self.load = load
        self._rr = 0
        self.predictor = predictor
        # Optional fleet-load view (gateway.cluster.FleetLoadView). When set, the
        # cost function scores by local + peers' in-flight, so a backend busy on
        # another replica is treated as busy here too. None -> single-replica.
        self.fleet = fleet

    def _candidates(self, backends: list[str]) -> list[str]:
        return backends

    def _inflight(self, backend_id: str) -> int:
        """Effective in-flight = this replica's + peers' (fleet-wide when clustered)."""
        n = self.load.inflight[backend_id]
        if self.fleet is not None:
            n += self.fleet.peer_inflight(backend_id)
        return n

    def choose(self, prompt: str, backends: list[str], strategy: str | None = None,
               seed: int = 0) -> RouteResult:
        strategy = strategy or self.cfg.strategy
        if self.cfg.use_bpe_hashing:
            hashes, mode = block_hashes_with_fallback(
                prompt, self.cfg.block_chars, self.cfg.block_tokens,
                self.cfg.hash_cutoff_blocks, self.cfg.tokenizer_model,
                use_bpe=True, seed=seed,
            )
        else:
            hashes = block_hashes(prompt, self.cfg.block_chars,
                                  self.cfg.hash_cutoff_blocks, seed=seed)
            mode = "char"
        tokens = len(hashes) * self.cfg.block_tokens
        cands = self._candidates(backends)

        if strategy == "round_robin":
            b = cands[self._rr % len(cands)]
            self._rr += 1
            return RouteResult(b, hashes, tokens, 0, mode)

        if strategy == "consistent_hash":
            key = hashes[0] if hashes else 0
            b = cands[key % len(cands)]
            return RouteResult(b, hashes, tokens, 0, mode)

        # prefix_tree: minimize estimated TTFT, then load-balance within hysteresis
        match = self.tree.match(hashes)
        scored = []                            # (ttft, backend, match_blocks)
        saturated_fallback = None
        for b in cands:
            # --- guardrails: hard saturation cutoff (affinity yields to load) ---
            # A backend a PEER has shed (circuit-open there) is treated as
            # saturated here too, so a fleet-wide-bad node is avoided everywhere.
            peer_bad = self.fleet is not None and self.fleet.peer_unhealthy(b)
            if peer_bad or self.load.kv_usage[b] > self.cfg.kv_pressure_cutoff or \
               self._inflight(b) > self.cfg.max_inflight:
                if saturated_fallback is None or \
                   self._inflight(b) < self._inflight(saturated_fallback):
                    saturated_fallback = b
                continue
            m_blocks = match.get(b, 0)
            scored.append((self._est_ttft(tokens, m_blocks, b), b, m_blocks))

        if not scored:                         # everything saturated
            b = saturated_fallback or cands[0]
            return RouteResult(b, hashes, tokens, match.get(b, 0), mode)

        # Backends whose TTFT is within hysteresis of the best are "equivalent":
        # spread across them by least-load (+ round-robin) so a tiny shared prefix
        # (e.g. a common system prompt) doesn't pin all traffic to one node, while
        # a large real affinity still keeps a single backend the sole winner.
        min_ttft = min(s[0] for s in scored)
        good = [(b, m) for (ttft, b, m) in scored if ttft <= min_ttft + self.cfg.hysteresis_ms]
        min_load = min(self._inflight(b) for b, _ in good)
        tied = [(b, m) for (b, m) in good if self._inflight(b) == min_load]
        b, m_blocks = tied[self._rr % len(tied)]
        self._rr += 1
        return RouteResult(b, hashes, tokens, m_blocks, mode)

    def _est_ttft(self, tokens: int, match_blocks: int, backend_id: str) -> float:
        cached_tokens = match_blocks * self.cfg.block_tokens
        uncached = max(0, tokens - cached_tokens)
        # MVP: linear prefill. Phase 2: a*(N-m)*N + b*(N-m) fitted from real timings.
        prefill_ms = self.cfg.prefill_ms_per_token * uncached
        queue_ms = self._inflight(backend_id) * self.cfg.service_ms_per_request
        static_ttft = prefill_ms + queue_ms
        if self.predictor is None or not self.predictor.available():
            return static_ttft
        feats = features_for(prompt_tokens=tokens, match_blocks=match_blocks,
                             block_tokens=self.cfg.block_tokens,
                             inflight=self._inflight(backend_id),
                             kv_usage=self.load.kv_usage[backend_id],
                             backend_id=backend_id)
        return blended_ttft(static_ttft, self.predictor.predict(feats),
                            self.cfg.ttft_predictor_weight)

File: rust/src/router.rs
//! Routing strategies (mirrors Python `gateway/router.py`).
//!
//! prefix_tree minimizes estimated TTFT, then spreads within `hysteresis_ms` of
//! the best by least-load + round-robin so a tiny shared prefix can't pin all
//! traffic to one node while a large real affinity still isolates one backend.

use crate::config::{Config, Strategy};
use crate::hashing::block_hashes;
use crate::load::LoadTracker;
use crate::radix_tree::RadixTree;

#[derive(Debug, Clone)]
pub struct RouteResult {
    pub backend_id: String,
    pub hashes: Vec<u64>,
    pub tokens: usize,
    pub match_blocks: usize,
}

pub struct Router {
    pub cfg: Config,
    rr: usize,
}

impl Router {
    pub fn new(cfg: Config) -> Self {
        Router { cfg, rr: 0 }
    }

    pub fn choose(
        &mut self,
        prompt: &str,
        backends: &[String],
        tree: &RadixTree,
        load: &LoadTracker,
        strategy: Strategy,
        seed: u64,
    ) -> RouteResult {
        let hashes = block_hashes(prompt, self.cfg.block_chars, self.cfg.hash_cutoff_blocks, seed);
        let tokens = hashes.len() * self.cfg.block_tokens;

        match strategy {
            Strategy::RoundRobin => {
                let b = backends[self.rr % backends.len()].clone();
                self.rr += 1;
                RouteResult { backend_id: b, hashes, tokens, match_blocks: 0 }
            }
            Strategy::ConsistentHash => {
                let key = hashes.first().copied().unwrap_or(0);
                let idx = (key % backends.len() as u64) as usize;
                RouteResult { backend_id: backends[idx].clone(), hashes, tokens, match_blocks: 0 }
            }
            Strategy::PrefixTree => {
                let matches = tree.match_prefix(&hashes);
                let mut scored: Vec<(f64, String, usize)> = Vec::new();
                let mut saturated_fallback: Option<String> = None;

                for b in backends {
                    if load.kv_usage(b) > self.cfg.kv_pressure_cutoff
                        || load.inflight(b) > self.cfg.max_inflight
                    {
                        match &saturated_fallback {
                            None => saturated_fallback = Some(b.clone()),
                            Some(sb) => {
                                if load.inflight(b) < load.inflight(sb) {
                                    saturated_fallback = Some(b.clone());
                                }
                            }
                        }
                        continue;
                    }
                    let mb = *matches.get(b).unwrap_or(&0);
                    scored.push((self.est_ttft(tokens, mb, b, load), b.clone(), mb));
                }

                if scored.is_empty() {
                    let b = saturated_fallback.unwrap_or_else(|| backends[0].clone());
                    let mb = *matches.get(&b).unwrap_or(&0);
                    return RouteResult { backend_id: b, hashes, tokens, match_blocks: mb };
                }

                let min_ttft = scored.iter().map(|s| s.0).fold(f64::INFINITY, f64::min);
                let good: Vec<(String, usize)> = scored
                    .iter()
                    .filter(|s| s.0 <= min_ttft + self.cfg.hysteresis_ms)
                    .map(|s| (s.1.clone(), s.2))
                    .collect();
                let min_load = good.iter().map(|(b, _)| load.inflight(b)).min().unwrap();
                let tied: Vec<(String, usize)> =
                    good.into_iter().filter(|(b, _)| load.inflight(b) == min_load).collect();
                let (b, mb) = tied[self.rr % tied.len()].clone();
                self.rr += 1;
                RouteResult { backend_id: b, hashes, tokens, match_blocks: mb }
            }
        }
    }

    fn est_ttft(&self, tokens: usize, match_blocks: usize, backend: &str, load: &LoadTracker) -> f64 {
        let cached = match_blocks * self.cfg.block_tokens;
        let uncached = tokens.saturating_sub(cached);
        let prefill = self.cfg.prefill_ms_per_token * uncached as f64;
        let queue = load.inflight(backend) as f64 * self.cfg.service_ms_per_request;
        prefill + queue
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::hashing::block_hashes;

    fn backends() -> Vec<String> {
        vec!["b0".into(), "b1".into(), "b2".into()]
    }

    #[test]
    fn round_robin_cycles() {
        let mut r = Router::new(Config::default());
        let tree = RadixTree::new(2000);
        let load = LoadTracker::new();
        let bs = backends();
        let seq: Vec<String> = (0..6)
            .map(|_| r.choose("p", &bs, &tree, &load, Strategy::RoundRobin, 0).backend_id)
            .collect();
        assert_eq!(seq, vec!["b0", "b1", "b2", "b0", "b1", "b2"]);
    }

    #[test]
    fn prefix_tree_prefers_cached_backend() {
        let cfg = Config::default();
        let mut tree = RadixTree::new(cfg.backend_cache_blocks);
        let load = LoadTracker::new();
        let prompt = "X".repeat(1300); // ~20 blocks -> affinity > hysteresis
        let h = block_hashes(&prompt, cfg.block_chars, cfg.hash_cutoff_blocks, 0);
        tree.insert(&h, "b1");
        let mut r = Router::new(cfg);
        let res = r.choose(&prompt, &backends(), &tree, &load, Strategy::PrefixTree, 0);
        assert_eq!(res.backend_id, "b1");
        assert_eq!(res.match_blocks, h.len());
    }

    #[test]
    fn prefix_tree_saturation_cutoff_excludes_overloaded() {
        let cfg = Config::default();
        let mut tree = RadixTree::new(cfg.backend_cache_blocks);
        let mut load = LoadTracker::new();
        let prompt = "X".repeat(1300);
        let h = block_hashes(&prompt, cfg.block_chars, cfg.hash_cutoff_blocks, 0);
        tree.insert(&h, "b1");
        load.set_inflight("b1", cfg.max_inflight + 1); // b1 has the cache but is saturated
        let mut r = Router::new(cfg);
        let res = r.choose(&prompt, &backends(), &tree, &load, Strategy::PrefixTree, 0);
        assert_ne!(res.backend_id, "b1");
    }
}
